zerver/lib/subscription_info.py

```diff
@@ -559,18 +569,35 @@ def bulk_get_subscriber_user_ids(
             zerver_subscription.user_profile_id
         FROM
             zerver_subscription
+        JOIN zerver_userprofile on zerver_userprofile.id = zerver_subscription.user_profile_id
         WHERE
-            zerver_subscription.recipient_id in %(recipient_ids)s AND
             zerver_subscription.active AND
-            zerver_subscription.is_user_active
+            zerver_subscription.is_user_active AND
+            (
+                zerver_subscription.recipient_id in %(full_fetch_recipient_ids)s
+                OR
+                (
+                    zerver_subscription.recipient_id in %(partial_fetch_recipient_ids)s AND
+                    (zerver_userprofile.is_bot OR (NOT zerver_userprofile.long_term_idle))
+                )
+            )
         ORDER BY
             zerver_subscription.recipient_id,
             zerver_subscription.user_profile_id
         """
     )
 
     cursor = connection.cursor()
-    cursor.execute(query, {"recipient_ids": tuple(recipient_ids)})
+    cursor.execute(
+        query,
+        {
+            "full_fetch_recipient_ids": tuple(full_fetch_recipient_ids),
+            # SQL doesn't allow "IN" with empty lists
```

Comment:
Is there a nicer way to handle this?

Reply:
Would it be better to have an `if` statement creating different queries depending on if `partial_fetch_recipient_ids` is empty? Oh I guess `full_fetch_recipient_ids` could be empty too. So that would be three branches of an if statement...?

Reply:
Three branches sounds kinda terrible. 

I think you can maybe just replace IN with the ANY operator to avoid that: https://www.postgresql.org/docs/current/functions-subquery.html#FUNCTIONS-SUBQUERY-ANY-SOME

Reply:
I couldn't get that working. Continued the conversation here: https://chat.zulip.org/#narrow/channel/3-backend/topic/partial.20subscriber.20data/with/2206670